geo/src/algorithm/line_intersection.rs

```diff
@@ -0,0 +1,666 @@
+        } else if p.end == q.start || p.end == q.end {
+            intersection = p.end;
+            // Now check to see if any endpoint lies on the interior of the other segment.
+        } else if p_q1 == Collinear {
+            intersection = q.start;
+        } else if p_q2 == Collinear {
+            intersection = q.end;
+        } else if q_p1 == Collinear {
+            intersection = p.start;
+        } else {
+            assert_eq!(q_p2, Collinear);
+            intersection = p.end;
+        }
+        Some(LineIntersection::SinglePoint {
+            intersection,
+            is_proper: false,
+        })
+    } else {
+        let intersection = proper_intersection(p, q);
+        Some(LineIntersection::SinglePoint {
+            intersection,
+            is_proper: true,
+        })
+    }
+}
+
+fn collinear_intersection<F: GeoFloat>(p: Line<F>, q: Line<F>) -> Option<LineIntersection<F>> {
+    let p_rect = p.bounding_rect();
+    let p1q1p2 = p_rect.intersects(&q.start);
+    let p1q2p2 = p_rect.intersects(&q.end);
+
+    let q_rect = q.bounding_rect();
+    let q1p1q2 = q_rect.intersects(&p.start);
+    let q1p2q2 = q_rect.intersects(&p.end);
+
+    if p1q1p2 && p1q2p2 {
+        return Some(LineIntersection::Collinear { intersection: q });
+    }
+    if q1p1q2 && q1p2q2 {
+        return Some(LineIntersection::Collinear { intersection: p });
+    }
+    if p1q1p2 && q1p1q2 {
+        if q.start == p.start && !p1q2p2 && !q1p2q2 {
+            return Some(LineIntersection::SinglePoint {
+                intersection: q.start,
+                is_proper: false,
+            });
+        } else {
+            return Some(LineIntersection::Collinear {
+                intersection: Line::new(q.start, p.start),
+            });
+        }
+    }
+    if p1q1p2 && q1p2q2 {
+        if q.start == p.end && !p1q2p2 && !q1p1q2 {
+            return Some(LineIntersection::SinglePoint {
+                intersection: q.start,
+                is_proper: false,
+            });
+        } else {
+            return Some(LineIntersection::Collinear {
+                intersection: Line::new(q.start, p.end),
+            });
+        }
+    }
+    if p1q2p2 && q1p1q2 {
+        if q.end == p.start && !p1q1p2 && !q1p2q2 {
+            return Some(LineIntersection::SinglePoint {
+                intersection: q.end,
+                is_proper: false,
+            });
+        } else {
+            return Some(LineIntersection::Collinear {
+                intersection: Line::new(q.end, p.start),
+            });
+        }
+    }
+    if p1q2p2 && q1p2q2 {
+        if q.end == p.end && !p1q1p2 && !q1p1q2 {
+            return Some(LineIntersection::SinglePoint {
+                intersection: q.end,
+                is_proper: false,
+            });
+        } else {
+            return Some(LineIntersection::Collinear {
+                intersection: Line::new(q.end, p.end),
+            });
+        }
+    }
```

Comment:
I played around with formatting these conditional checks a different way:

```rust
    Some(match (p1q1p2, p1q2p2, q1p1q2, q1p2q2) {
        (true, true, _, _) => LineIntersection::Collinear { intersection: q },
        (_, _, true, true) => LineIntersection::Collinear { intersection: q },
        (true, false, true, false) if q.start == p.start => {
            LineIntersection::SinglePoint {
                intersection: q.start,
                is_proper: false,
            }
        }
        (true, _, true, _) => {
            LineIntersection::Collinear {
                intersection: Line::new(q.start, p.start),
            }
        }
        ...
        _ => return None,
    })
```

As always, take it or leave it. Was just experimenting

Reply:
It's easier to see the patterns when laid out as pattern matching -- good call. I've done that (and introduced a couple helper functions for brevity).